File: src/models/pokemon.js
'use strict';

const Bluebird = require('bluebird');

const Bookshelf           = require('../libraries/bookshelf');
const Box                 = require('./box');
const Evolution           = require('./evolution');
const GameFamily          = require('./game-family');
const GameFamilyDexNumber = require('./game-family-dex-number');
const Location            = require('./location');

module.exports = Bookshelf.model('Pokemon', Bookshelf.Model.extend({
  tableName: 'pokemon',
  hasTimestamps: ['date_created', 'date_modified'],
  game_family () {
    return this.belongsTo(GameFamily, 'game_family_id');
  },
  game_family_dex_numbers () {
    return this.hasMany(GameFamilyDexNumber, 'pokemon_id');
  },
  boxes () {
    return this.hasMany(Box, 'pokemon_id');
  },
  locations () {
    return this.hasMany(Location, 'pokemon_id');
  },
  box (query) {
    if (query.game_family === undefined || query.regional === undefined) {
      return null;
    }

    const box = this.related('boxes').models.find((m) => {
      return m.get('game_family_id') === query.game_family && m.get('regional') === query.regional;
    });

    if (!box) {
      return null;
    }

    return box.get('value');
  },
  capture_summary (query) {
    return Object.assign({
      id: this.get('id'),
      national_id: this.get('national_id'),
      name: this.get('name'),
      game_family: this.related('game_family').serialize(),
      form: this.get('form'),
      box: this.box(query)
    }, this.get('dex_number_properties'));
  },
  evolutions (query) {
    return new Evolution()
    .where('evolutions.evolution_family_id', this.get('evolution_family_id'))
    .query((qb) => {
      qb.joinRaw('INNER JOIN pokemon AS evolved ON evolutions.evolved_pokemon_id = evolved.id');
      qb.joinRaw('INNER JOIN pokemon AS evolving ON evolutions.evolving_pokemon_id = evolving.id');
      qb.joinRaw('INNER JOIN game_families AS evolved_game_family ON evolved.game_family_id = evolved_game_family.id');
      qb.joinRaw('INNER JOIN game_families AS evolving_game_family ON evolving.game_family_id = evolving_game_family.id');

      if (query.game_family) {
        qb.whereRaw(`
          evolved_game_family.order <= (
            SELECT "order" FROM game_families WHERE id = ?
          ) AND
          evolving_game_family.order <= (
            SELECT "order" FROM game_families WHERE id = ?
          )
        `, [query.game_family, query.game_family]);
      }
      if (query.regional) {
        qb.joinRaw('LEFT OUTER JOIN game_family_dex_numbers AS evolved_dex_numbers ON evolved.id = evolved_dex_numbers.pokemon_id');
        qb.joinRaw('LEFT OUTER JOIN game_family_dex_numbers AS evolving_dex_numbers ON evolving.id = evolving_dex_numbers.pokemon_id');
        qb.whereRaw(`evolved_dex_numbers.game_family_id = ? AND evolving_dex_numbers.game_family_id = ?`, [query.game_family, query.game_family]);
      }

      qb.orderByRaw('CASE WHEN trigger = \'breed\' THEN evolving.national_id ELSE evolved.national_id END, trigger DESC, evolved.national_order ASC');
    })
    .fetchAll({ withRelated: Evolution.RELATED })
    .get('models');
  },
  virtuals: {
    dex_number_properties () {
      return this.related('game_family_dex_numbers')
        .reduce((dexNumbers, dexNumber) => {
          const numbers = Object.assign({}, dexNumbers);
          numbers[`${dexNumber.get('game_family_id')}_id`] = dexNumber.get('dex_number');

          return numbers;
        }, {});
    },
    summary () {
      return {
        id: this.get('id'),
        national_id: this.get('national_id'),
        name: this.get('name'),
        form: this.get('form')
      };
    }
  },
  serialize (request) {
    const query = request.query || {};

    return this.evolutions(query)
    .reduce((family, evolution) => {
      const i = evolution.get('stage') - 1;
      const breed = evolution.get('trigger') === 'breed';
      let first;
      let second;

      family.pokemon[i] = family.pokemon[i] || [];
      family.pokemon[i + 1] = family.pokemon[i + 1] || [];
      if (breed) {
        first = evolution.related('evolved_pokemon').get('summary');
        second = evolution.related('evolving_pokemon').get('summary');
      } else {
        first = evolution.related('evolving_pokemon').get('summary');
        second = evolution.related('evolved_pokemon').get('summary');
      }

      if (!family.pokemon[i].find((p) => p.id === first.id)) {
        family.pokemon[i].push(first);
      }
      if (!family.pokemon[i + 1].find((p) => p.id === second.id)) {
        family.pokemon[i + 1].push(second);
      }

      family.evolutions[i] = family.evolutions[i] || [];
      family.evolutions[i].push(evolution.serialize());

      return family;
    }, { pokemon: [], evolutions: [] })
    .then((family) => {
      // filter out nulls from evolutions that don't exist in the given game
      // family or regionality
      while (family.pokemon.length > 0 && !family.pokemon[0]) {
        family.pokemon.shift();
      }
      while (family.evolutions.length > 0 && !family.evolutions[0]) {
        family.evolutions.shift();
      }

      if (family.pokemon.length === 0) {
        family.pokemon.push([this.get('summary')]);
      }
      return Bluebird.all([
        family,
        query.game_family && new GameFamily({ id: query.game_family }).fetch({ require: true })
      ]);
    })
    .spread((evolutionFamily, gameFamily) => {
      const locations = this.related('locations')
        .filter((l) => {
          if (!gameFamily) {
            return true;
          }

          const locationGameFamily = l.related('game').related('game_family');

          if (query.regional) {
            return gameFamily.id === locationGameFamily.get('id');
          }

          return gameFamily.get('generation') >= locationGameFamily.get('generation');
        })
        .map((l) => l.serialize(request));

      return Object.assign({
        id: this.get('id'),
        national_id: this.get('national_id'),
        name: this.get('name'),
        game_family: this.related('game_family').serialize(),
        form: this.get('form'),
        box: this.box(query)
      }, this.get('dex_number_properties'), {
        locations,
        evolution_family: evolutionFamily
      });
    });
  }
}, {
  CAPTURE_SUMMARY_RELATED: ['boxes', 'game_family', 'game_family_dex_numbers'],
  RELATED: ['boxes', 'game_family', 'game_family_dex_numbers', {
    locations (qb) {
      qb
        .innerJoin('games', 'locations.game_id', 'games.id')
        .innerJoin('game_families', 'games.game_family_id', 'game_families.id')
        .orderByRaw('game_families.order DESC, games.order ASC');
    }
  }, 'locations.game', 'locations.game.game_family']
}));

File: pkg/models/pokemon.go
package models

import (
	"strings"

	"github.com/go-pg/pg/orm"
)

type Pokemon struct {
	tableName struct{} `sql:"pokemon,alias:pokemon"`

	ID           int        `json:"id"`
	NationalID   int        `json:"national_id"`
	Name         string     `json:"name"`
	GameFamilyID string     `json:"-"`
	GameFamily   GameFamily `json:"game_family"`
	Form         *string    `json:"form"`
	Box          *string    `json:"box"`

	RedBlueID                *int `sql:"-" json:"red_blue_id,omitempty"`
	YellowID                 *int `sql:"-" json:"yellow_id,omitempty"`
	GoldSilverID             *int `sql:"-" json:"gold_silver_id,omitempty"`
	CrystalID                *int `sql:"-" json:"crystal_id,omitempty"`
	RubySapphireID           *int `sql:"-" json:"ruby_sapphire_id,omitempty"`
	FireRedLeafGreenID       *int `sql:"-" json:"fire_red_leaf_green_id,omitempty"`
	EmeraldID                *int `sql:"-" json:"emerald_id,omitempty"`
	DiamondPearlID           *int `sql:"-" json:"diamond_pearl_id,omitempty"`
	PlatinumID               *int `sql:"-" json:"platinum_id,omitempty"`
	HeartGoldSoulSilverID    *int `sql:"-" json:"heart_gold_soul_silver_id,omitempty"`
	BlackWhiteID             *int `sql:"-" json:"black_white_id,omitempty"`
	Black2White2ID           *int `sql:"-" json:"black_2_white_2_id,omitempty"`
	XYID                     *int `sql:"-" json:"x_y_id,omitempty"`
	OmegaRubyAlphaSapphireID *int `sql:"-" json:"omega_ruby_alpha_sapphire_id,omitempty"`
	SunMoonID                *int `sql:"-" json:"sun_moon_id,omitempty"`
	UltraSunUltraMoonID      *int `sql:"-" json:"ultra_sun_ultra_moon_id,omitempty"`

	XLocation    string `json:"-"`
	YLocation    string `json:"-"`
	ORLocation   string `json:"-"`
	ASLocation   string `json:"-"`
	SunLocation  string `json:"-"`
	MoonLocation string `json:"-"`
	USLocation   string `json:"-"`
	UMLocation   string `json:"-"`

	XLocations    []string `sql:"-" json:"x_locations"`
	YLocations    []string `sql:"-" json:"y_locations"`
	ORLocations   []string `sql:"-" json:"or_locations"`
	ASLocations   []string `sql:"-" json:"as_locations"`
	SunLocations  []string `sql:"-" json:"sun_locations"`
	MoonLocations []string `sql:"-" json:"moon_locations"`
	USLocations   []string `sql:"-" json:"us_locations"`
	UMLocations   []string `sql:"-" json:"um_locations"`

	EvolutionFamilyID int             `json:"-"`
	EvolutionFamily   EvolutionFamily `sql:"-" json:"evolution_family"`
}

func (p *Pokemon) AfterQuery(db orm.DB) error {
	p.XLocations = splitLocations(p.XLocation)
	p.YLocations = splitLocations(p.YLocation)
	p.ORLocations = splitLocations(p.ORLocation)
	p.ASLocations = splitLocations(p.ASLocation)
	p.SunLocations = splitLocations(p.SunLocation)
	p.MoonLocations = splitLocations(p.MoonLocation)
	p.USLocations = splitLocations(p.USLocation)
	p.UMLocations = splitLocations(p.UMLocation)

	return nil
}

func (p *Pokemon) LoadDexNumbers(dexNumbers []*GameFamilyDexNumber) {
	for _, dn := range dexNumbers {
		if dn.PokemonID != p.ID {
			continue
		}
		n := dn.DexNumber
		switch dn.GameFamilyID {
		case RedBlueID:
			p.RedBlueID = &n
		case YellowID:
			p.YellowID = &n
		case GoldSilverID:
			p.GoldSilverID = &n
		case CrystalID:
			p.CrystalID = &n
		case RubySapphireID:
			p.RubySapphireID = &n
		case FireRedLeafGreenID:
			p.FireRedLeafGreenID = &n
		case EmeraldID:
			p.EmeraldID = &n
		case DiamondPearlID:
			p.DiamondPearlID = &n
		case PlatinumID:
			p.PlatinumID = &n
		case HeartGoldSoulSilverID:
			p.HeartGoldSoulSilverID = &n
		case BlackWhiteID:
			p.BlackWhiteID = &n
		case Black2White2ID:
			p.Black2White2ID = &n
		case XYID:
			p.XYID = &n
		case OmegaRubyAlphaSapphireID:
			p.OmegaRubyAlphaSapphireID = &n
		case SunMoonID:
			p.SunMoonID = &n
		case UltraSunUltraMoonID:
			p.UltraSunUltraMoonID = &dn.DexNumber
		}
	}
}

func (p *Pokemon) LoadEvolutions(evolutions []*Evolution) {
	p.EvolutionFamily = EvolutionFamily{[][]PokemonSummary{}, [][]Evolution{}}
	for _, e := range evolutions {
		if e.EvolutionFamilyID != p.EvolutionFamilyID {
			continue
		}

		i := e.Stage - 1
		breed := e.Trigger == "breed"

		var first, second PokemonSummary

		if breed {
			first, second = e.EvolvedPokemon, e.EvolvingPokemon
		} else {
			first, second = e.EvolvingPokemon, e.EvolvedPokemon
		}

		if len(p.EvolutionFamily.Pokemon) <= i {
			p.EvolutionFamily.Pokemon = append(p.EvolutionFamily.Pokemon, []PokemonSummary{})
		}
		if len(p.EvolutionFamily.Pokemon) <= i+1 && len(p.EvolutionFamily.Pokemon) < 3 {
			p.EvolutionFamily.Pokemon = append(p.EvolutionFamily.Pokemon, []PokemonSummary{})
		}

		if !findPokemon(p.EvolutionFamily.Pokemon[i], first) {
			p.EvolutionFamily.Pokemon[i] = append(p.EvolutionFamily.Pokemon[i], first)
		}
		if !findPokemon(p.EvolutionFamily.Pokemon[i+1], second) {
			p.EvolutionFamily.Pokemon[i+1] = append(p.EvolutionFamily.Pokemon[i+1], second)
		}

		if len(p.EvolutionFamily.Evolutions) <= i {
			p.EvolutionFamily.Evolutions = append(p.EvolutionFamily.Evolutions, []Evolution{})
		}
		p.EvolutionFamily.Evolutions[i] = append(p.EvolutionFamily.Evolutions[i], *e)
	}

	// for pokemon without any evolutions, add itself to the Pokemon slice
	if len(p.EvolutionFamily.Pokemon) == 0 {
		summary := PokemonSummary{
			ID:         p.ID,
			NationalID: p.NationalID,
			Name:       p.Name,
			Form:       p.Form,
		}
		p.EvolutionFamily.Pokemon = append(p.EvolutionFamily.Pokemon, []PokemonSummary{summary})
	}
}

type EvolutionFamily struct {
	Pokemon    [][]PokemonSummary `json:"pokemon"`
	Evolutions [][]Evolution      `json:"evolutions"`
}

type PokemonSummary struct {
	tableName struct{} `sql:"pokemon,alias:pokemon"`

	ID         int     `json:"id"`
	NationalID int     `json:"national_id"`
	Name       string  `json:"name"`
	Form       *string `json:"form"`
}

type Evolution struct {
	tableName struct{} `sql:"evolutions,alias:evolutions"`

	EvolvingPokemonID int            `sql:",pk" json:"-"`
	EvolvedPokemonID  int            `sql:",pk" json:"-"`
	EvolvingPokemon   PokemonSummary `json:"-"`
	EvolvedPokemon    PokemonSummary `json:"-"`
	EvolutionFamilyID int            `json:"-"`
	Stage             int            `json:"-"`
	Trigger           string         `json:"trigger"`
	Level             *int           `json:"level,omitempty"`
	Stone             *string        `json:"stone,omitempty"`
	HeldItem          *string        `json:"held_item,omitempty"`
	Notes             *string        `json:"notes,omitempty"`
}

func splitLocations(location string) []string {
	if location != "" {
		return strings.Split(location, ", ")
	}
	return []string{}
}

func findPokemon(summaries []PokemonSummary, summary PokemonSummary) bool {
	for _, s := range summaries {
		if s.ID == summary.ID {
			return true
		}
	}
	return false
}
